lib/auth/index.js

```diff
@@ -0,0 +1,277 @@
+                    await deleteUserToken('_')
+                } else {
+                    setupRefreshTimer('_')
+                }
+            }
+        } catch (err) {
+            RED.log.error('FlowFuse Assistant: Failed to load access tokens')
+        }
+    }
+}
+
+/**
+ * Write tokens to file
+ */
+async function saveTokens () {
+    await writeFile(tokenFile, JSON.stringify(activeTokens), 'utf8')
+}
+
+/**
+ * Update a user token, setup refresh timer and save to file
+ * @param {*} user
+ * @param {*} token
+ */
+async function setUserToken (user, token) {
+    activeTokens[user] = token
+    setupRefreshTimer(user)
+    await saveTokens()
+}
+
+/**
+ * Setup the refresh timer for the token. User tokens have a relatively short expiry time, but can be refreshed before expiry
+ * to provide a longer session.
+ * @param {*} user
+ */
+function setupRefreshTimer (user) {
+    const token = activeTokens[user]
+    if (!token.expires_at) {
+        token.expires_at = Date.now() + (token.expires_in * 1000)
+    }
+    const refreshInterval = token.expires_at - Date.now() - 10000 // refresh with 10 secs to spare
+    if (refreshInterval > 0) {
+        activeTimers[user] = setTimeout(async () => {
+            try {
+                const newTokens = await refreshToken(token)
+                newTokens.expires_at = Date.now() + (newTokens.expires_in * 1000)
+                await setUserToken(user, newTokens)
+            } catch (err) {
+                // Failed to refresh token - remove it and disable the agent
+                await deleteUserToken(user)
+                assistant.RED.log.warn('Failed to refresh FlowFuse Assistant access token')
+                reinitialiseAssistant()
+            }
+        }, refreshInterval)
+    }
+}
+
+/**
+ * When running in a managed FF environment, the tokens are provided statically via settings.
+ * This is used to insert the token into the auth table so we have a single place to retrieve
+ * tokens from.
+ * @param {*} token
+ */
+function setStaticToken (token) {
+    // A token provided by settings file - does not need refresh logic
+    activeTokens._ = {
+        access_token: token
+    }
+}
+
+/**
+ * Get the current user token
+ */
+function getUserToken () {
+    // Only single shared user supported currently
+    return activeTokens._?.access_token
+}
+
+/**
+ * Delete a users token, clears up refresh timers and updates saved tokens
+ * @param {*} user
+ */
+async function deleteUserToken (user) {
+    const token = activeTokens[user]
+    if (token) {
+        clearTimeout(activeTimers[user])
+    }
+    delete activeTokens[user]
+    delete activeTimers[user]
+    saveTokens()
+}
+
+/**
+ * Make an API request to refresh the user token
+ * @param {*} token
+ */
+async function refreshToken (token) {
+    const params = {
+        grant_type: 'refresh_token',
+        client_id: 'ff-plugin',
```

Comment:
Any benefit in setting a unique (and identifiable) client ID like `ff-plugin:assistant` ?